### PR TITLE
add yolov5l

### PR DESCRIPTION
**Pull request by qingkwl**
add yolov5L

### CODE REVIEW

**Comment by qingkwl** [comment]
# YOLOv5 🚀 by Ultralytics, GPL-3.0 license
# Hyperparameters for high-augmentation COCO training from scratch
# python train.py --batch 32 --cfg yolov5m6.yaml --weights '' --data coco.yaml --img 1280 --epochs 300
# See tutorials for hyperparameter evolution https://github.com/ultralytics/yolov5#tutorials

lr0: 0.01  # initial learning rate (SGD=1E-2, Adam=1E-3)
lrf: 0.1  # final OneCycleLR learning rate (lr0 * lrf)
momentum: 0.937  # SGD momentum/Adam beta1
weight_decay: 0.0005  # optimizer weight decay 5e-4
warmup_epochs: 3.0  # warmup epochs (fractions ok)
warmup_momentum: 0.8  # warmup initial momentum
warmup_bias_lr: 0.1  # warmup initial bias lr
box: 0.05  # box loss gain
cls: 0.3  # cls loss gain
cls_pw: 1.0  # cls BCELoss positive_weight
obj: 0.7  # obj loss gain (scale with pixels)
obj_pw: 1.0  # obj BCELoss positive_weight
iou_t: 0.20  # IoU training threshold
anchor_t: 4.0  # anchor-multiple threshold
# anchors: 3  # anchors per output layer (0 to ignore)
fl_gamma: 0.0  # focal loss gamma (efficientDet default gamma=1.5)
hsv_h: 0.015  # image HSV-Hue augmentation (fraction)
hsv_s: 0.7  # image HSV-Saturation augmentation (fraction)
hsv_v: 0.4  # image HSV-Value augmentation (fraction)
degrees: 0.0  # image rotation (+/- deg)
translate: 0.1  # image translation (+/- fraction)
scale: 0.9  # image scale (+/- gain)
shear: 0.0  # image shear (+/- deg)
perspective: 0.0  # image perspective (+/- fraction), range 0-0.001
flipud: 0.0  # image flip up-down (probability)
fliplr: 0.5  # image flip left-right (probability)
mosaic: 1.0  # image mosaic (probability)
mixup: 0.1  # image mixup (probability)
copy_paste: 0.1  # segment copy-paste (probability)
paste_in: 0.0  # image copy paste (probability), use 0 for faster training
max_box_per_img: 160 # label mask

**Comment by qingkwl** [comment]
from mindspore import ops
from mindspore import nn
from mindspore import boost
from mindspore import context
from mindspore.ops import functional as F
from mindspore.ops import composite as C
from mindspore.common import dtype as mstype
from mindspore.nn.wrap.cell_wrapper import TrainOneStepCell
from mindspore.boost.boost_cell_wrapper import BoostTrainOneStepCell
from mindspore.nn.wrap.cell_wrapper import _TrainPipelineAccuStepCell, _pipeline_clear_grad
from mindspore.nn.wrap.loss_scale import _TrainPipelineWithLossScaleCell
from mindspore.train.amp import validator, _check_level, _check_kwargs, _config_level, \
    _do_keep_batchnorm_fp32, auto_mixed_precision, _add_loss_network, _get_pipeline_stages

GRADIENT_CLIP_TYPE = 1
GRADIENT_CLIP_VALUE = 10.0
clip_grad = C.MultitypeFuncGraph("clip_grad")


@clip_grad.register("Number", "Number", "Tensor")
def _clip_grad(clip_type, clip_value, grad):
    """
    Clip gradients.

    Inputs:
        clip_type (int): The way to clip, 0 for 'value', 1 for 'norm'.
        clip_value (float): Specifies how much to clip.
        grad (tuple[Tensor]): Gradients.

    Outputs:
        tuple[Tensor], clipped gradients.
    """
    if clip_type not in (0, 1):
        return grad
    dt = F.dtype(grad)
    if clip_type == 0:
        new_grad = C.clip_by_value(grad, F.cast(F.tuple_to_array((-clip_value,)), dt),
                                   F.cast(F.tuple_to_array((clip_value,)), dt))
    else:
        new_grad = nn.ClipByNorm()(grad, F.cast(F.tuple_to_array((clip_value,)), dt))
    return new_grad

**Comment by qingkwl** [comment]
class _TrainOneStepCell(TrainOneStepCell):
    def __init__(self, network, ema, optimizer, amp_loss_scaler, sens=1.0, enable_clip_grad=True):
        super(_TrainOneStepCell, self).__init__(network, optimizer, sens=sens)
        self.ema = ema
        self.use_loss_scaler = False if amp_loss_scaler is None else True
        self.amp_loss_scaler = amp_loss_scaler
        self.enable_clip_grad = enable_clip_grad
        print(f"[INFO] Enable loss scale: {self.use_loss_scaler}", flush=True)

    def construct(self, *inputs):
        loss = self.network(*inputs)
        sens = F.fill(loss.dtype, loss.shape, self.sens)
        grads = self.grad(self.network, self.weights)(*inputs, sens)
        grads = self.grad_reducer(grads)
        if self.use_loss_scaler:
            grads = self.amp_loss_scaler.unscale(grads)
        if self.enable_clip_grad:
            grads = self.hyper_map(F.partial(clip_grad, GRADIENT_CLIP_TYPE, GRADIENT_CLIP_VALUE), grads)
        loss = F.depend(loss, self.optimizer(grads))
        loss = F.depend(loss, self.ema.update())
        return loss

**Comment by qingkwl** [comment]
class _BoostTrainOneStepCell(BoostTrainOneStepCell):
    def __init__(self, network, ema, optimizer, amp_loss_scaler, sens=1.0, enable_clip_grad=True):
        super(_BoostTrainOneStepCell, self).__init__(network, optimizer, sens)
        self.ema = ema
        self.use_loss_scaler = False if amp_loss_scaler is None else True
        self.amp_loss_scaler = amp_loss_scaler
        self.enable_clip_grad = enable_clip_grad
        print(f"[INFO] Enable loss scale: {self.use_loss_scaler}", flush=True)

    def construct(self, *inputs):
        if self.freeze:
            loss = self.gradient_freeze_process(*inputs)
        else:
            loss = self.network(*inputs)
            sens = F.fill(loss.dtype, loss.shape, self.sens)
            grads = self.grad(self.network, self.weights)(*inputs, sens)
            grads = self.grad_reducer(grads)
            if self.use_loss_scaler:
                grads = self.amp_loss_scaler.unscale(grads)
            if self.use_grad_accumulation:
                loss = self.gradient_accumulation_process(loss, grads, sens, *inputs)
            else:
                if self.enable_dim_reduce:
                    loss = F.depend(loss, self.dim_reduce(loss, grads, sens, self.weights, self.weights_clone, *inputs))
                elif self.enable_adasum:
                    loss = F.depend(loss, self.adasum_process(loss, grads))
                else:
                    if self.enable_clip_grad:
                        grads = self.hyper_map(F.partial(clip_grad, GRADIENT_CLIP_TYPE, GRADIENT_CLIP_VALUE), grads)
                    loss = F.depend(loss, self.optimizer(grads))

        loss = F.depend(loss, self.ema.update())
        return loss

**Comment by qingkwl** [comment]
def initialize_weights(model):
    for _, m in model.cells_and_names():
        classname = m.__class__.__name__
        if classname == "Conv2d":
            pass  # nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
        elif classname in ("BatchNorm2d", "SyncBatchNorm"):
            m.eps = 1e-3
            m.momentum = 0.03
